src/LSST/system/TelemetryItem.h

```diff
@@ -114,115 +132,297 @@ class TelemetryItem {
+    /// Set the value at `index` in `_vals` to val.
+    /// @return false if index is out of range.
+    bool setVal(size_t index, VT val) {
+        if (index > _size) {
+            return false;
+        }
+        std::lock_guard<std::mutex> lg(_mtx);
+        _vals[index] = val;
+        return true;
+    }
 
-    /// Return the value of `_commCurrent`.
-    double getCommCurrent() { return _commCurrent->getVal(); }
+    //// Return the value at `index` in `_vals`.
+    /// @throw TelemetryException if index is out of range.
+    VT getVal(size_t index) const {
+        if (index > _size) {
+            throw TelemetryException(ERR_LOC, "TItemVector::getVal out of range for index=" +
+                                     std::to_string(index) + " for " + dump());
+        }
+        std::lock_guard<std::mutex> lg(_mtx);
+        return _vals[index];
+    }
 
-    /// Local override of getJson
-    nlohmann::json getJson() const override { return buildJsonFromMap(_map); }
+    /// Return the json representation of this object.
+    nlohmann::json getJson() const override {
+        nlohmann::json js;
+        nlohmann::json jArray = nlohmann::json::array();
+
+        std::lock_guard<std::mutex> lg(_mtx);
+        for (size_t j = 0; j < _size; ++j) {
+            jArray.push_back(_vals[j]);
+        }
+        js[getId()] = jArray;
+        return js;
+    }
 
     /// Set the value of this object from json.
     bool setFromJson(nlohmann::json const& js, bool idExpected) override {
-        return setMapFromJson(_map, js, idExpected);
+        if (idExpected) {
+            // This type can only have a floating point value array for `_val`.
+            LERROR("TItemVector::setFromJson cannot have a json 'id' entry");
+            return false;
+        }
+        try {
+            std::vector<VT> vals = js.at(getId());
+            return setVals(vals);
+        } catch (nlohmann::json::out_of_range const& ex) {
+            LERROR("TItemVector::setFromJson out of range for ", getId(), " js=", js);
+        }
+        return false;
     }
 
     /// Return true if this item and `other` have the same id and values.
-    bool compareItem(TelemetryItem const& other) const override;
+    bool compareItem(TelemetryItem const& other) const override {
+        try {
+            TItemVector const& otherT = dynamic_cast<TItemVector const&>(other);
+            if (getId() != other.getId() || _size != otherT._size) {
+                return false;
+            }
+            // There's no way to reliably control which gets locked first
+            // (both A.compareItem(B) or B.compareItem(A) are possible),
+            // so this is needed to lock both without risk of deadlock.
+            std::unique_lock lockThis(_mtx, std::defer_lock);
+            std::unique_lock lockOther(otherT._mtx, std::defer_lock);
+            std::lock(lockThis, lockOther);
+            return (_vals == otherT._vals);
```

Comment:
The requirement document needs to have 20 Hz telemetry frequency with a very small bandwidth (I know the vendor never reaches this). If you have `mutex` lock for the telemetry, the overhead might hurt the performance in the future. But we can consider this after the code works.

Reply:
I don't expect very many blind comparisons during operations. This is primarily used in unit tests, but since it is there, it should be safe to call.